packages/core/src/anchor/ethereum/ethereum-anchor-service.ts

```diff
@@ -97,179 +95,216 @@ export default class EthereumAnchorService extends AnchorService {
-
-                this.emit(cidDoc.docId, failedRes)
-                return // exit loop
-            }
-
-            await new Promise(resolve => setTimeout(resolve, DEFAULT_POLL_TIME))
-
-            try {
-                const requestUrl = [this.requestsApiEndpoint, cidDoc.cid.toString()].join('/')
-                const response = await fetch(requestUrl)
-
-                if (response.status === HTTP_STATUS_NOT_FOUND) {
-                    // the anchor request does not exist, fail and stop polling
-                    // TODO
-                    this.emit(cidDoc.docId, {
-                        cid: cidDoc.cid, status: AnchorStatus[AnchorStatus.FAILED], message: 'Request not found.'
-                    })
-                    poll = false
-                    break
-                }
-
-                const json = await response.json()
-                switch (json.status) {
-                    case "PENDING": {
-                        // just log
-                        break
-                    }
-                    case "PROCESSING": {
-                        this.emit(cidDoc.docId, { cid: cidDoc.cid, status: json.status, message: json.message })
-                        break
-                    }
-                    case "FAILED": {
-                        this.emit(cidDoc.docId, { cid: cidDoc.cid, status: json.status, message: json.message })
-                        poll = false
-                        break
-                    }
-                    case "COMPLETED": {
-                        const { anchorRecord } = json
-                        const anchorRecordCid = new CID(anchorRecord.cid.toString())
-
-                        this.emit(cidDoc.docId, {
-                            cid: cidDoc.cid, status: json.status, message: json.message, anchorRecord: anchorRecordCid
-                        })
-                        poll = false
-                        break
-                    }
-                }
-            } catch (e) {
-                // just log
-            }
+      })
+    );
+  }
+
+  /**
+   * Start polling for CidDocPair mapping
+   * @param cidDoc - CID to Doc mapping
+   * @param pollTime - Single poll timeout
+   * @param maxPollingTime - Global timeout for max polling in milliseconds
+   * @private
+   */
+  private poll(
+    cidDoc: CidDoc,
+    pollTime?: number,
+    maxPollingTime?: number
+  ): Observable<AnchorServiceResponse> {
+    const started = new Date().getTime();
+    const maxTime = started + (maxPollingTime | DEFAULT_MAX_POLL_TIME);
+    const requestUrl = [this.requestsApiEndpoint, cidDoc.cid.toString()].join(
+      "/"
+    );
+
+    return interval(DEFAULT_POLL_TIME).pipe(
+      concatMap(async () => {
+        const now = new Date().getTime();
+        if (now > maxTime) {
+          throw new Error("Exceeded max timeout");
+        } else {
+          const response = await fetch(requestUrl);
+          if (response.status === HTTP_STATUS_NOT_FOUND) {
```

Comment:
hmm, I wonder if we should be handling all failed statuses here, not just `HTTP_STATUS_NOT_FOUND`...

Probably out of scope for this change though